plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchange.java

```diff
@@ -0,0 +1,290 @@
+            partitions.forEach(partitionFilesMap::put);
+        }
+
+        ImmutableList.Builder<ExchangeSourceHandle> result = ImmutableList.builder();
+        for (Integer partitionId : partitionFilesMap.keySet()) {
+            result.add(new FileSystemExchangeSourceHandle(partitionId, ImmutableList.copyOf(partitionFilesMap.get(partitionId)), secretKey));
+        }
+        return result.build();
+    }
+
+    private URI getCommittedAttemptPath(FileSystemExchangeSinkHandle sinkHandle)
+    {
+        URI sinkOutputBasePath = getExchangeDirectory()
+                .resolve(sinkHandle.getTaskPartitionId() + PATH_SEPARATOR);
+        try {
+            List<URI> attemptPaths = exchangeStorage.listDirectories(sinkOutputBasePath).collect(toImmutableList());
+            checkState(!attemptPaths.isEmpty(), "no attempts found for sink %s", sinkHandle);
+
+            List<URI> committedAttemptPaths = attemptPaths.stream()
+                    .filter(this::isCommitted)
+                    .collect(toImmutableList());
+            checkState(!committedAttemptPaths.isEmpty(), "no committed attempts found for %s", sinkHandle);
+
+            return committedAttemptPaths.get(0);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean isCommitted(URI attemptPath)
+    {
+        URI commitMarkerFilePath = attemptPath.resolve(COMMITTED_MARKER_FILE_NAME);
+        return exchangeStorage.exists(commitMarkerFilePath);
+    }
+
+    private Map<Integer, URI> getCommittedPartitions(URI committedAttemptPath)
+    {
+        try {
+            List<URI> partitionFiles = exchangeStorage.listFiles(committedAttemptPath)
+                    .filter(file -> file.getPath().endsWith(DATA_FILE_SUFFIX))
+                    .collect(toImmutableList());
+            ImmutableMap.Builder<Integer, URI> result = ImmutableMap.builder();
+            for (URI partitionFile : partitionFiles) {
+                Matcher matcher = PARTITION_FILE_NAME_PATTERN.matcher(new File(partitionFile.getPath()).getName());
+                checkState(matcher.matches(), "unexpected partition file: %s", partitionFile);
+                int partitionId = Integer.parseInt(matcher.group(1));
+                result.put(partitionId, partitionFile);
+            }
+            return result.build();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private URI getExchangeDirectory()
+    {
+        return baseDirectory.resolve(exchangeContext.getQueryId() + "." + exchangeContext.getStageId() + PATH_SEPARATOR);
+    }
+
+    @Override
+    public CompletableFuture<List<ExchangeSourceHandle>> getSourceHandles()
+    {
+        return exchangeSourceHandlesFuture;
+    }
+
+    @Override
+    public ExchangeSourceSplitter split(ExchangeSourceHandle handle, long targetSizeInBytes)
+    {
+        // TODO: replace with correct split implementation
```

Comment:
QQ: do we still need this splitter? As we discussed, `StageTaskSourceFactory` will always split the files according to `targetSizeInBytes`. Does that mean we don't need this splitter anymore?

Reply:
IIUC the current implementation will produce one split per file. If the files are tiny it might not be necessary to produce as many partitions as there are files.